console/src/App/AppContextProvider.js

```diff
@@ -0,0 +1,86 @@
+import React, {Component} from 'react';
+import produce from "immer";
+
+const Context = React.createContext();
+
+class AppContextProvider extends Component {
+	constructor(props){
+		super(props)
+		this.state = {
+			userId: 1,
+			userSettings: {
+				color: '#FFFFFF',
+				fontSize: 0,
+				skin: props.defaultSkin || 'carbon',
+			},
+			updateAppState: this.updateAppState,
+			onSwitchUser: this.onSwitchUser
+		}
+	}
+
+	componentDidMount(){
+		this.loadUserSettings(this.state.userId);
+	}
+
+	loadUserSettings = (userId) => {
+		if(!JSON.parse(window.localStorage.getItem(`user${this.state.userId}`))){
+			window.localStorage.setItem(`user${this.state.userId}`, JSON.stringify({...this.state.userSettings}));
+		}
+
+		this.setState(
+			produce((draft) => {
+				draft.userSettings = JSON.parse(window.localStorage.getItem(`user${userId}`))
+			})
+		)
+	}
+
+	saveUserSettingsLocally = (userId, hasChanged) => {
+		if(hasChanged){
+			window.localStorage.setItem(`user${userId}`, JSON.stringify(this.state.userSettings))
+		}
+	}
+
+	// Catch all way to update state
+	updateAppState = (cb) => {
+		const prevUserSettings = this.state.userSettings
+		this.setState(
+			produce(cb),
+			() => {
+				this.saveUserSettingsLocally(this.state.userId, this.state.userSettings !== prevUserSettings)
```

Comment:
Why did you choose to use the `setState` callback rather than `componentDidUpdate`?